epios/sampler_age.py

```diff
@@ -0,0 +1,221 @@
+from epios.sampler import Sampler
+import numpy as np
+import json
+
+
+class SamplerAge(Sampler):
+    '''
+    The sampling class with age stratification.
+
+    Parameters:
+    -----------
+
+    If you want to input new data, you can input that into data argument and set the pre_process to True
+    If you want to use previous processed data, you can input the data_store_path to read data files,
+    and set the pre_process to False.
+
+    num_age_group : int
+        Indicating how many age groups are there.
+
+        *The last group includes age >= some threshold*
+    age_group_width : int
+        Indicating the width of each age group(except for the last group)
+    mode : str
+        This indicates the specific mode to process the data.
+        This should be the name of the modes that can be identified.
+
+        **If you want this class sample as originally designed, do not change this value**
+
+    '''
+
+    def __init__(self, data=None, data_store_path='./input/', pre_process=True, num_age_group=17, age_group_width=5,
+                 mode='Age'):
+        self.mode = mode
+        super().__init__(data=data, data_store_path=data_store_path,
+                         num_age_group=num_age_group, pre_process=pre_process,
+                         age_group_width=age_group_width, mode=self.mode)
+        ageinfo_path = data_store_path + 'pop_dist.json'
+        self.ageinfo = ageinfo_path
+        self.age_group_width = age_group_width
+
+    def get_age_dist(self):
+        '''
+        Read the age distribution from pop_dist.json processed from DataProcess class
+
+        Output:
+        -------
+
+        config : list
+            A list of floats, with sum 1, length should be the number of age groups
+
+        '''
+        with open(self.ageinfo, 'r') as f:
+            config = json.loads(f.read())
+        return config
+
+    def bool_exceed(self, current_age: int, cap_age: float):
+        '''
+        Return a boolean value to tell whether the sampling is going to exceed any cap
+        --------
+        Input:
+        All inputs should be integers
+
+        Output:
+        True - means not reaching the cap
+        False - means reaching the cap
+
+        '''
+        if current_age + 2 > cap_age:
+            return False
+        else:
+            return True
+
+    def multinomial_draw(self, n: int, prob: list):
+        '''
+        Perform a multinomial draw with caps, it will return a tuple of lists.
```

Comment:
Does the numpy.random multinomial sampling function not do this?

Reply:
Here we want to set a cap for each age group, where the numpy.random multinomial function cannot let us to set the caps for each group. Also, setting up the caps in a 2D table may have some mathematical issues, also need to deal with that.